apollo-federation/src/operation/merging.rs

```diff
@@ -28,43 +29,38 @@ impl<'a> FieldSelectionValue<'a> {
     /// Returns an error if:
     /// - The parent type or schema of any selection does not match `self`'s.
     /// - Any selection does not select the same field position as `self`.
-    fn merge_into<'op>(
-        &mut self,
-        others: impl Iterator<Item = &'op FieldSelection>,
-    ) -> Result<(), FederationError> {
+    fn merge_into(&mut self, other: &FieldSelection) -> Result<(), FederationError> {
         let self_field = &self.get().field;
-        let mut selection_sets = vec![];
-        for other in others {
-            let other_field = &other.field;
-            if other_field.schema != self_field.schema {
-                return Err(FederationError::internal(
-                    "Cannot merge field selections from different schemas",
-                ));
-            }
-            if other_field.field_position != self_field.field_position {
-                return Err(FederationError::internal(format!(
+        let mut selection_set = None;
+        let other_field = &other.field;
+        if other_field.schema != self_field.schema {
+            return Err(FederationError::internal(
+                "Cannot merge field selections from different schemas",
+            ));
+        }
+        if other_field.field_position != self_field.field_position {
+            return Err(FederationError::internal(format!(
                     "Cannot merge field selection for field \"{}\" into a field selection for field \"{}\"",
                     other_field.field_position,
                     self_field.field_position,
                 )));
-            }
-            if self.get().selection_set.is_some() {
-                let Some(other_selection_set) = &other.selection_set else {
-                    return Err(FederationError::internal(format!(
-                        "Field \"{}\" has composite type but not a selection set",
-                        other_field.field_position,
-                    )));
-                };
-                selection_sets.push(other_selection_set);
-            } else if other.selection_set.is_some() {
+        }
+        if self.get().selection_set.is_some() {
+            let Some(other_selection_set) = &other.selection_set else {
                 return Err(FederationError::internal(format!(
                     "Field \"{}\" has non-composite type but also has a selection set",
                     other_field.field_position,
                 )));
-            }
+            };
+            selection_set = Some(other_selection_set);
+        } else if other.selection_set.is_some() {
+            return Err(FederationError::internal(format!(
+                "Field \"{}\" has non-composite type but also has a selection set",
+                other_field.field_position,
+            )));
```

Comment:
looks like we flipped error messages on those conditions?

Reply:
Ah, ya. I made that mistake during clean up...